Add formation state builder for Opta

MKlaasman opened the pull request:
This Pull Request relates to issue 124: https://github.com/PySport/kloppy/issues/124 

First of all, I am quite new to committing to an open-source repo, so feel free to give pointers.

As stated in the issue, the goal of this Pull Request is to add an extra state to the dataset. In this case specifically for the Opta deserializer. 

Similar to the score state, home_score and away_score, we first initialize the beginning formations from the F7 files. Using a FormationStateBuilder, we change the formation for either the home or away team when a “FormationChangeEvent” occurs (Opta event type_id=”40”). 

Multiple extra classes are added to support this functionality, namely: 
-	Formation (common.py)
-	FormationType (models\event.py)
-	FormationChangeEvent (models\event.py)
-	FormationStateBuilder (services\state_builder\builders\formation.py)

In this branch, I also added the “formation_names” dictionary (for mapping formation_ids to formations) to the opta/deserializer.py, however, I was not sure whether this should live in a separate config file?

I did run tests/opta.py, which was successful for the Opta files that live in the Kloppy repository. However, these files do not contain any formation changes. Therefore, I tested the code out for different Opta files, which showed expected results. 

`
events = dataset.to_pandas(    
    additional_columns={
        “home_formation”: lambda event: event.state[“formation”].home,
        “away_formation”: lambda event: event.state[“formation”].away,
    }
}
`

Example of a section of the corresponding output event:
![example_kloppy_formation_state_output](https://user-images.githubusercontent.com/59951106/148086362-a9fafa64-9a9f-4a13-a0f5-c916f5d7b77e.png)


Before committing I reformatted using “black .”, if for any reason I forgot anything, please let me know. 
Note: Still need to update Documentation? (If approved)

koenvo commented on kloppy/domain/models/event.py:
```diff
@@ -130,6 +130,17 @@ class CardType(Enum):
+        FORMATION_TYPE (FormationType): formation the team changes to
+    """
+
+    FORMATION_TYPE = None
```
This should include the list of formations. This list: https://github.com/PySport/kloppy/pull/125/files#diff-6b63d1c70bf1c1cdc45872bdf3bebb33bb5b9696761ffff0f16e66d2a5538fe5R194

koenvo commented on kloppy/infra/serializers/event/opta/deserializer.py:
```diff
@@ -186,6 +191,33 @@
     77: "player off pitch",
 }
 
+formation_names = {
```
This should be a mapping from `qualifier_id (int)` to `FormationType`

koenvo commented on kloppy/domain/models/common.py:
```diff
@@ -21,6 +21,20 @@ class Score:
+        away:
+    """
+
+    home: str
```
the `home` and `away` attributes should be of type `FormationType`

koenvo replied:
Second though. I think it would be good to move this to `Team.starting_formation`

koenvo commented on kloppy/domain/models/common.py:
```diff
@@ -670,6 +685,7 @@ class Metadata:
     periods: List[Period]
     pitch_dimensions: PitchDimensions
     score: Score
+    formations: Formation
```
This should move to `Team.starting_formation`

koenvo commented on kloppy/domain/models/event.py:
```diff
@@ -568,6 +581,24 @@ class CardEvent(Event):
+    Attributes:
+        event_type (EventType): `EventType.FORMATION_CHANGE` (See [`EventType`][kloppy.domain.models.event.EventType])
+        event_name (str): `"card"`
+        card_type: See [`CardType`][kloppy.domain.models.event.CardType]
```
Please update this `card_type` attribute to `formation`

koenvo commented on kloppy/domain/services/state_builder/builders/formation.py:
```diff
@@ -0,0 +1,37 @@
+
+@dataclass
+class Formation:
+    home: str
```
`home` and `away` should be of type `FormationType`

koenvo commented:
This looks great! I added some comments which are mostly about consistent usage of the `FormationType`. Can you also add a unit test for both opta FormationChange event and for the state_builder? If you need some assistance please let me know, happy to help!

MKlaasman commented:
Thanks for the feedback @koenvo. I hope this is what you envisioned, if not please reach out. 

-	**Added the same usability to the Statsbomb deserializer**, since I wanted to double-check whether I didn’t implement anything Opta specific into the codebase. So updated possible FormationTypes to include Statsbomb possibilities. No Documentation was found on all the possible formations. Therefore, I had contact with a Statsbomb employee and eventually decided to parse all these (https://github.com/statsbomb/open-data/tree/master/data/events) events from Statsbomb open data to get a list of possible formations. Therefore quite confident but not certain that all formations are within the formation mapping for StatsBomb and FormationType.

-	**All tests pass. Added multiple tests to:**
   •	test_opta.py (Team.starting_formation), 
  •	test_statsbomb.py (Team.starting_formation), and 
  •	test_state_builder.py (FormationChangeEvent, Team.starting_formation, event.state[“formation”])

-	There were no formation changes within the corresponding opta file within tests/files, therefore I use the “statsbomb_event.json” to build the corresponding formation state test within test_state_builder.py 

-	By moving starting_formation to Team. I could retire the Formation Class from common.py, since we won't need a combined Formation class, like we have the Score class. Class Formation still exists within the formation state builder

-	Added an empty starting_formation as default to the Team class. So other provider won’t break.

•	Formation state_builder and FormationChangeEvent won’t be usable for other data providers than Opta and Statsbomb. Note: might want to include this in Documentation.

Let me know if I can add or remove anything,

Cheers,
Milan

koenvo commented:
Thanks for taking your time to work on this PR!

There might be some conflicts due to the merge of @JanVanHaaren I just merged. Let me merge your PR tomorrow and solve those issues right away.
